[Tools] Normalizing mri_protocol: split up comma-separated values

llevitis opened the pull request:
This script serves as the first step to normalizing the mri_protocol. Some projects store comma-separated values for multiple fields, so `normalize_protocol_split_rows` creates arrays of unique combinations from each row and inserts new rows with unique values. It then deletes the original row containing comma-separated values. The subsequent step to fully normalizing the mri_protocol will be a 2nd PR containing a patch to replace all %_range columns with %_min & %_max columns and a script to populate the new columns. 

IMPORTANT NOTE: this should not be merged unless there are PR on the LORIS-MRI side to account for these changes and ready to be merged.

cmadjar commented:
@driusan @llevitis @MounaSafiHarab I added the blocked and needs work labels for the same reason as the other mri_protocol normalization PR. As long as this does not come with a ready to merge PR on the LORIS-MRI side, it should not be merged.

MounaSafiHarab commented on tools/normalize_protocol_split_rows.php:
```diff
@@ -0,0 +1,73 @@
+}
+
+// Get the list of unique IDs from the mri_protocol_3_3 table
+$mp_idx = $DB->pselect("SELECT ID FROM mri_protocol_3", array());
```
I don't think we have a table called mri_protocol_3 :)

MounaSafiHarab commented on tools/normalize_protocol_split_rows.php:
```diff
@@ -0,0 +1,73 @@
+    $DB->delete("mri_protocol", array("ID" => $id));
+}
+
+echo("All mri_protocol entries are now unique and not comma-separated.\n\n");
```
not critical, but it might be useful for the user running this script to have a few interactive messages as to which rows are being split and into what values 
(I ran the script and it runs fine, but for me, I have NO values separated by commas in my table, so it might have been nice to say: No entry have comma separated values, or something like that).

MounaSafiHarab replied:
I faked an entry with commas into my existing, and here is what I get:

Before:
![before](https://user-images.githubusercontent.com/15198379/37924604-5ab0e3f0-3100-11e8-915c-5c48c2906ce0.png)


After:
![after](https://user-images.githubusercontent.com/15198379/37924614-60b1e876-3100-11e8-893c-9e591247a421.png)

seems like your `cartesian` function is working!

MounaSafiHarab replied:
@cmadjar @nicolasbrossard 

from what I know, the ID column in the mri_protocol table is unused on the Loris-MRI side, can you confirm too? 
because @llevitis's script here deletes entries and regenerates new ones, resulting in new auto-incremented IDs.
Should not be an issue, but just confirming with you.

cmadjar replied:
Indeed, I don't believe the ID column in the mri_protocol table is used so it should be ok.

nicolasbrossard commented on tools/normalize_protocol_split_rows.php:
```diff
@@ -0,0 +1,73 @@
+}
+
+// Get the list of unique IDs from the mri_protocol table
+$mp_idx = $DB->pselect("SELECT ID FROM mri_protocol", array());
```
Since you are getting the whole table anyway on line 56, I would refactor this like so:

```php
$rows = $DB->pselect("SELECT ID FROM mri_protocol", array());

foreach ($rows as $row) {
    foreach($row as $key => $value) {
    .
    .
    .
```
and replace `$id` on line 67 with `$value['ID']`. What do you think?

llevitis commented:
@MounaSafiHarab - I added a check for whether any commas have been detected in an existing row of the table. If none are detected, the script skips to the next row so as to avoid unnecessary deletion and redundant insertion. A statement is also echoed indicating that no commas were detected and the table is thus unaltered. 

@nicolasbrossard - Good point! I have refactored the code as you suggested.

ridz1208 commented:
@MounaSafiHarab changes after you have tested, please test again

@nicolasbrossard please dissmiss or re-review

ridz1208 commented:
@cmadjar removed the needs work as this PR could be ready, blocked remains as per your comments above

nicolasbrossard commented:
Code looks good. I tested the final version and it works for me. Approving changes.

cmadjar commented:
@ridz1208 The MRI PR needs to be updated and properly tested. As long as it is not ready to be merged on the LORIS-MRI side, I don't think we should unblock this one.

@MounaSafiHarab what do you think?

ridz1208 commented:
@cmadjar okai

Still blocked, Loris-MRI PR not ready

MounaSafiHarab commented on tools/normalize_protocol_split_rows.php:
```diff
@@ -0,0 +1,85 @@
+        $DB->delete($table_name, array("ID" => $id));
+    }
+    if ($total_commas == 0) {
+        echo("No commas have been detected. The $table_name table has been unaltered.\n\n");
```
Just a minor detail;
when the user runs this script; this message, together with the message below; both get outputted to the screen which is a bit confusing (see screenshot):
![messagemisleading](https://user-images.githubusercontent.com/15198379/40498674-a3ac4db2-5f4d-11e8-898d-d5c12c1395ba.png)

MounaSafiHarab commented:
> Still blocked, Loris-MRI PR not ready

@cmadjar the MRI one is ready now, if you want to unblock this?

driusan commented:
@llevitis @cmadjar @MounaSafiHarab Any word on the required MRI-side changes so that this can be merged?

MounaSafiHarab commented:
> @llevitis @cmadjar @MounaSafiHarab Any word on the required MRI-side changes so that this can be merged?

@driusan 
they should be ready, but since I sent the PR, I will wait for @cmadjar and/or @nicolasbrossard to merge the one on the MRI side (just tagged them).

MounaSafiHarab commented:
@driusan 
the MRI one is already merged by @cmadjar 
please feel free to merge this one as well
